core/src/main/java/io/questdb/cairo/arr/FunctionArray.java

```diff
@@ -0,0 +1,182 @@
+                }
+                break;
+            case ColumnType.SHORT:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putShort(functions[i].getShort(record));
+                }
+                break;
+            case ColumnType.INT:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putInt(functions[i].getInt(record));
+                }
+                break;
+            case ColumnType.LONG:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putLong(functions[i].getLong(record));
+                }
+                break;
+            case ColumnType.DATE:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putLong(functions[i].getDate(record));
+                }
+                break;
+            case ColumnType.TIMESTAMP:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putLong(functions[i].getTimestamp(record));
+                }
+                break;
+            case ColumnType.FLOAT:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putFloat(functions[i].getFloat(record));
+                }
+                break;
+            case ColumnType.DOUBLE:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putDouble(functions[i].getDouble(record));
+                }
+                break;
+            case ColumnType.LONG256:
+                for (int i = 0; i < elemCount; i++) {
+                    Long256 v = functions[i].getLong256A(record);
+                    mem.putLong256(v.getLong0(), v.getLong1(), v.getLong2(), v.getLong3());
+                }
+                break;
+            case ColumnType.UUID:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putLong128(functions[i].getLong128Lo(record), functions[i].getLong128Hi(record));
+                }
+                break;
+            case ColumnType.IPv4:
+                for (int i = 0; i < elemCount; i++) {
+                    mem.putInt(functions[i].getIPv4(record));
+                }
+                break;
+            default:
+                throw new AssertionError("impossible array element type");
+        }
+    }
+
+    public void applyShape(CairoConfiguration configuration, int errorPos) {
+        resetToDefaultStrides(configuration.maxArrayElementCount(), errorPos);
+        if (functions == null || functions.length < flatViewLength) {
+            functions = new Function[flatViewLength];
+        }
+    }
+
+    @Override
+    public void close() {
+        Function[] functions = functions();
+        for (int n = functions.length, i = 0; i < n; i++) {
+            functions[i] = Misc.free(functions[i]);
+        }
+    }
+
+    @Override
+    public double getDoubleAtAbsIndex(int flatIndex) {
+        validateFlatIndex(flatIndex);
+        return functions[flatIndex].getDouble(record);
```

Comment:
The purpose of the `functions()` method is questionable, but it's used in all other methods here:
```suggestion
        return functions()[flatIndex].getDouble(record);
```